atomic_reactor/plugins/pre_add_image_content_manifest.py

```diff
@@ -0,0 +1,208 @@
+    minimal_icm = {
+        'metadata': {
+            'icm_version': 1,
+            'icm_spec': ('https://raw.githubusercontent.com/containerbuildsystem/atomic-reactor/'
+                         'master/atomic_reactor/schemas/content_manifest.json'),
+            'image_layer_index': -1,
+        },
+        'content_sets': [],
+        'image_contents': [],
+    }
+
+    def __init__(self, tasker, workflow, remote_source_icm_url=None, destdir=IMAGE_BUILD_INFO_DIR):
+        """
+        :param tasker: ContainerTasker instance
+        :param workflow: DockerBuildWorkflow instance
+        :param icm_url: str, URL of the ICM from the Cachito request.
+        :param destdir: image path to carry content_manifests data dir
+        """
+        super(AddImageContentManifestPlugin, self).__init__(tasker, workflow)
+        self.content_manifests_dir = os.path.join(destdir, 'content_manifests', '')
+        self.icm_url = remote_source_icm_url
+        self.dfp = df_parser(self.workflow.builder.df_path, workflow=self.workflow)
+        labels = Labels(self.dfp.labels)
+        _, image_name = labels.get_name_and_value(Labels.LABEL_TYPE_COMPONENT)
+        _, image_version = labels.get_name_and_value(Labels.LABEL_TYPE_VERSION)
+        _, image_release = labels.get_name_and_value(Labels.LABEL_TYPE_RELEASE)
+        self.icm_file_name = '{}-{}-{}.json'.format(image_name, image_version, image_release)
+        self.content_sets = []
+        self._cachito_verify = True
+        self._layer_index = None
+        self._icm = None
+
+    @property
+    def layer_index(self):
+        if not self._layer_index:
+            # Default layer index is 1, because base and 'FROM scratch' images
+            #     *always* have 2 layers
+            self._layer_index = 1
+        if not base_image_is_scratch(self.dfp.baseimage):
+            inspect = self.workflow.builder.base_image_inspect
+            self._layer_index = len(inspect[INSPECT_ROOTFS][INSPECT_ROOTFS_LAYERS])
+        return self._layer_index
+
+    @property
+    def cachito_verify(self):
+        if self._cachito_verify is True:
+            cachito_conf = get_cachito(self.workflow) or {}
+            # Get the value of Cachito's 'insecure' key from the active reactor config map,
+            #    *flip it*, and let the result tell us whether to verify or not
+            self._cachito_verify = not cachito_conf.get('insecure', False)
+        return self._cachito_verify
+
+    @property
+    def icm(self):
+        """
+        Get and validate the ICM from the Cachito API `content-manifest` endpoint.
+
+        :return: dict, the ICM as a Python dict
+        """
+        if self.icm_url is None:
+            self._icm = self.minimal_icm
+        if self._icm is None:
+            session = get_retrying_requests_session()
+            session.verify = self.cachito_verify
+            self.log.debug('Making request to "%s"', self.icm_url)
+            response = session.get(self.icm_url)
```

Comment:
Don't we have `CachitoAPI` class that should implement all cachito operations?

Reply:
We do, but the URL comes from the API response itself and then forwarded here so we do not need to add an additional dependency on the cachito util module (see resolve_remote_sources). We do the same for the configuration files in the download remote sources plugin.